### PR TITLE
Refactor DataX.Flow to support batch scenarios better

### PR DESCRIPTION
**Pull request by kjcho-msft**
We don't need to generate Transform, Projection and Schema files when there is no job to deploy so the following will be executed after GenerateJobConfig and GenerateJobConfigBatch.
GenerateTransformFile
GenerateProjectionFile
GenerateSchemaFile

### CODE REVIEW

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+            var runtimeKeyVaultName = flowToDeploy.GetTokenString(PortConfigurationSettings.TokenName_RuntimeKeyVaultName);
+            Ensure.NotNull(runtimeKeyVaultName, "runtimeKeyVaultName");
+
+            var secretName = $"{config.Name}-transform";
+            Configuration.TryGet(Constants.ConfigSettingName_SparkType, out string sparkType);
+            var uriPrefix = (sparkType != null && sparkType == Constants.SparkTypeDataBricks) ? Constants.PrefixSecretScope : Constants.PrefixKeyVault;
+
+            var transformFileSecret = $"{uriPrefix}://{runtimeKeyVaultName}/{secretName}";
+            flowToDeploy.SetStringToken(TokenName_TransformFile, transformFileSecret);
```
Comment (vijayupadya) [inline Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: Where is this secret saved to keyvault? In general it is not a good practice to have the secret token set without saving the secret. If the secret save fails, we will have a non-existent key set in the token.

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs [hunk 2]
```diff
@@ -153,9 +153,9 @@ public static async Task<List<string>> GetLastModifiedBlobContentsInBlobPath(str
             CloudBlobClient cloudBlobClient = storageAccount.CreateCloudBlobClient();
             CloudBlobContainer container = cloudBlobClient.GetContainerReference(containerName);
 
-            var filteredBlobs = container.ListBlobsSegmentedAsync(prefix: prefix, useFlatBlobListing: true, blobListingDetails: BlobListingDetails.None, maxResults: null, currentToken: null, options: null, operationContext: null).Result.Results.OfType<CloudBlockBlob>()
+            var filteredBlobs = await Task.Run(() => container.ListBlobsSegmentedAsync(prefix: prefix, useFlatBlobListing: true, blobListingDetails: BlobListingDetails.None, maxResults: null, currentToken: null, options: null, operationContext: null).Result.Results.OfType<CloudBlockBlob>()
```
Comment (vijayupadya) [inline Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: Why this change? i.e invoking the listBlobs in a separate task?

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S600_GenerateJobConfigBatch.cs [hunk 3]
```diff
@@ -224,7 +224,7 @@ private async Task<JobConfig> ScheduleSingleJob(JobDeploymentSession job, string
             destFolder = GetJobConfigFilePath(isOneTime, dateString, destFolder);
             var jc = new JobConfig
             {
-                Content = await GetBatchConfigContent(job, defaultJobConfig.ToString(), processStartTime, processEndTime).ConfigureAwait(false),
+                Content = GetBatchConfigContent(job, defaultJobConfig.ToString(), processStartTime, processEndTime),
```
Comment (vijayupadya) [inline Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S600_GenerateJobConfigBatch.cs]: Why async is removed?

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs (hunk 1 above)
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: Saving this secret to KV does happen in S610_GenerateTransformFile.cs. I made it on purpose. since if we don't have any job to deploy, we shouldn't save any file and even any secret.

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S600_GenerateJobConfigBatch.cs (hunk 3 above)
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S600_GenerateJobConfigBatch.cs]: I realized that it doesn't need to be async call.

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs (hunk 2 above)
Comment (kjcho-msft) [reply Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: If there are a lot of blobs in the target folder, it could probably affect the performance. Having this would help. Any concern? I agree it can be non-async call though.

File: Services/DataX.Config/DataX.Config.DatabricksClient/DatabricksClient.cs [hunk 4]
```diff
@@ -156,6 +156,7 @@ public async Task<SparkJobSyncResult> StopJob(JToken jobClientData)
 
         public async Task<SparkJobSyncResult[]> GetJobs()
         {
+            await Task.CompletedTask;
             throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config.DatabricksClient/DatabricksClient.cs]: Actually the code in this class was not part of my changes. This was added by Rohit. This is just a fix for bug 956780. @rohit489 can clarify this.

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+using DataX.Config.ConfigDataModel;
+
+namespace DataX.Config.ConfigGeneration.Processor
+{
+    /// <summary>
+    /// Produce the transform/query file section
+    /// </summary>
+    [Shared]
+    [Export(typeof(IFlowDeploymentProcessor))]
+    public class PrepareTransformFile : ProcessorBase
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: Existing config E2E tests are good enough, i mean, checking the job config output is good enough. Will add a comment like this. 
        /// <summary>
        /// Generate and set the info for the transform file which will be needed to generate JobConfig
        /// </summary>

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs (hunk 1 above)
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: And the reason why I have to compose this secret url here is that the url is needed to generate the job config which is s600.

Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: updated to use the secret url which was saved from the prepare steps.

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs (hunk 2 above)
Comment (kjcho-msft) [reply Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: update not to use task.run just use await.

File: Services/DataX.Config/DataX.Config.DatabricksClient/DatabricksClient.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohit489) [reply Services/DataX.Config/DataX.Config.DatabricksClient/DatabricksClient.cs]: We never call into GetJobs hence this was not implemented. However GetJobs is part of the ISparkJobClient hence it needed to be added to DatabricksClient class. In future if we ever require a GetJobs functionality then this method can be implemented at that time

[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config.DatabricksClient/DatabricksClient.cs]: Thanks, Rohit!

File: Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs [hunk 6]
```diff
@@ -51,26 +54,25 @@ public override async Task<string> Process(FlowDeploymentSession flowToDeploy)
...
-
             var runtimeKeyVaultName = flowToDeploy.GetTokenString(PortConfigurationSettings.TokenName_RuntimeKeyVaultName);
             Ensure.NotNull(runtimeKeyVaultName, "runtimeKeyVaultName");
 
-            var filePath = ResourcePathUtil.Combine(runtimeConfigBaseFolder, $"{config.Name}-combined.txt");
-            var transformFilePath = await RuntimeStorage.SaveFile(filePath, rulesCode.Code);
             var secretName = $"{config.Name}-transform";
-            var transformFileSecret = await KeyVaultClient.SaveSecretAsync(runtimeKeyVaultName, secretName, transformFilePath, Configuration.TryGet(Constants.ConfigSettingName_SparkType, out string sparkType) ? sparkType : null);
+            Configuration.TryGet(Constants.ConfigSettingName_SparkType, out string sparkType);
+            var uriPrefix = (sparkType != null && sparkType == Constants.SparkTypeDataBricks) ? Constants.PrefixSecretScope : Constants.PrefixKeyVault;
```
Comment (vijayupadya) [inline Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: This logic to compute uriPrefix is now in multiple places. Can we consolidate by adding a helper method to KeyVaultClient and call it from everywhere its needed including in KeyVaultClient. Only KeyVaultClient should know the nuances of uri based on sparkType.

Comment (kjcho-msft) [reply Services/DataX.Config/DataX.Config/ConfigGeneration/Processor/S450_PrepareTransformFile.cs]: Makes sense. Created a helper and updated to use it.
